File: Demos/FSharp/Funktionen.fs
module Funktionen

let hallo name = $"Hallo {name}!"
let hallo2 (name : string) = $"Hallo {name}!"

// int -> int -> int
let add a b = a + b

let add2 a = fun b -> a + b

let add10 = add 10

// ('a*'b -> 'c) -> 'a -> 'b -> 'c
let curry f a b = f (a,b)

// ('a*'b -> 'c) -> 'a -> 'b -> 'c
let partialApply f a =
   fun b -> f (a,b)

let add'(a,b) = a+b

let add10' = 
   partialApply add' 10

let add10'' = 
   curry add' 10

// string * string -> unit
let printName (punct, name) =
   printfn "Hallo %s%s" name punct

let printNameExcl =
   partialApply printName "!"

let pipe =
   1
   |> add 2 
   |> add 3

let inline srtpAdd a b =
   a + b

let inline trim (s : ^s when ^s : (member Trim : unit -> ^s)) =
   (^s : (member Trim : unit -> ^s) s)

File: Demos/CSharp/Funktionen.cs
using System;

static class Funktionen
{
   public static string Hallo(string name)
     => $"Hallo {name}!";

   public static int AddNotCurried(int a, int b)
     => a + b;
   public static Func<int,int> AddCurried(int a)
     => b => a + b;

   public static Func<int, int> Add10
      = AddCurried(10);

   public static Func<int,int> Add10alt1
      => FunExtensions.PartialApply<int,int,int>(AddNotCurried, 10);

   public static Func<int,int> Add10alt2
      => FunExtensions.Curry<int,int,int>(AddNotCurried)(10);

   public static void PrintName(string punct, string name)
      => Console.WriteLine($"Hallo {name}{punct}");

   public static Action<string> PrintNameExcl(string name)
      => FunExtensions.PartialApply<string,string>(PrintName, "!");

   public static int Add(this int a, int b)
     => a + b;

   public static int Pipe =
      1.Add(2).Add(3);
}

public static class FunExtensions
{
   public static Func<T1,Func<T2,T3>> Curry<T1,T2,T3>(Func<T1,T2,T3> f)
      => v1 => v2 => f(v1,v2);
   public static Func<T2,T3> PartialApply<T1,T2,T3>(Func<T1,T2,T3> f, T1 val1)
      => val2 => f(val1, val2);

   public static Action<T2> PartialApply<T1,T2>(Action<T1,T2> f, T1 val1)
      => val2 => f(val1, val2);

}